staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -348,41 +339,51 @@ func (req *request) Finish(execFn func()) bool {
 
 func (req *request) wait() (bool, bool) {
 	qs := req.qs
-	qs.lock.Lock()
-	defer qs.lock.Unlock()
-	if req.waitStarted {
-		// This can not happen, because the client is forbidden to
-		// call Wait twice on the same request
-		panic(fmt.Sprintf("Multiple calls to the Wait method, QueueSet=%s, startTime=%s, descr1=%#+v, descr2=%#+v", req.qs.qCfg.Name, req.startTime, req.descr1, req.descr2))
-	}
-	req.waitStarted = true
 
 	// ========================================================================
-	// Step 4:
+	// Step 3:
 	// The final step is to wait on a decision from
 	// somewhere and then act on it.
 	decisionAny := req.decision.Get()
-	qs.syncTimeLocked()
+	qs.lockAndSyncTime()
+	defer qs.lock.Unlock()
+	duplicateWait := req.waitStarted
+	if duplicateWait {
+		// This can not happen, because the client is forbidden to
+		// call Wait twice on the same request
+		klog.Errorf("Duplicate call to the Wait method!  Will return execute=false.  QueueSet=%s, startTime=%s, descr1=%#+v, descr2=%#+v", req.qs.qCfg.Name, req.startTime, req.descr1, req.descr2)
+	}
+	req.waitStarted = true
 	decision, isDecision := decisionAny.(requestDecision)
 	if !isDecision {
-		panic(fmt.Sprintf("QS(%s): Impossible decision %#+v (of type %T) for request %#+v %#+v", qs.qCfg.Name, decisionAny, decisionAny, req.descr1, req.descr2))
+		klog.Errorf("QS(%s): Impossible decision (type %T, value %#+v) for request %#+v %#+v!  Coercing to cancel", qs.qCfg.Name, decisionAny, decisionAny, req.descr1, req.descr2)
+		decision = decisionCancel
+	}
+	if duplicateWait {
+		return false, qs.isIdleLocked()
```

Comment:
instead of this - let's return in line 354 - it would be cleaner and this would be no-op change to put it there.

Reply:
Done.
It is a behavioral change, but one with no clear right answer in an impossible situtation.